tests/diffDataParent.js

```diff
@@ -0,0 +1,95 @@
+require([
+    'chai',
+    'jquery',
+    'underscore',
+    'tests/utils',
+    'text!static/diffDataParent/parse.xml'
+], function (
+    chai,
+    $,
+    _,
+    util,
+    PARSE_XML
+) {
+    var assert = chai.assert,
+        call = util.call;
+
+    describe("Control elements with different data parents", function() {
+        before(function (done) {
+            util.init({
+                javaRosa: {langs: ['en', 'hin']},
+                core: {onReady: done}
+            });
+        });
+
+        it("should not allow repeat group children to have other data parents", function() {
+            util.loadXML("");
+            var repeat = util.addQuestion("Repeat", 'repeat1'),
+                text1 = util.addQuestion.bind({prevId: repeat.p.nodeID})("Text", 'text1'),
+                text2 = util.addQuestion.bind({prevId: repeat.p.nodeID})("Text", 'text2');
+
+            text2.p.dataParent = '/data';
+            assert(util.isTreeNodeValid(text1), "text1 should be valid");
+            assert(!util.isTreeNodeValid(text2), "text2 should not be valid");
+        });
+
+        it("should not allow a data parent to be an input", function() {
+            util.loadXML("");
+            var text1 = util.addQuestion("Text", 'text1'),
+                text2 = util.addQuestion("Text", 'text2');
+
+            text2.p.dataParent = '/data/text1';
+            assert(util.isTreeNodeValid(text1), "text1 should be valid");
+            assert(!util.isTreeNodeValid(text2), "text2 should not be valid");
+        });
+
+        it("should keep the same data parent after a move", function() {
+            util.loadXML("");
+            var text1 = util.addQuestion("Text", 'text1'),
+                text2 = util.addQuestion("Text", 'text2'),
+                form = call("getData").core.form;
+
+            util.addQuestion("Group", 'group1');
+            text2.p.dataParent = '/data/group1';
+
+            form.moveMug(text2, text1, 'before');
+
+            assert.equal(text2.p.dataParent, '/data/group1');
+        });
+
+        it("should update data tree after a change to data parent", function() {
+            util.loadXML("");
+            var text1 = util.addQuestion("Text", 'text1'),
+                group1 = util.addQuestion("Group", 'group1'),
+                group2 = util.addQuestion.bind({prevId: text1.p.nodeID})("Group", 'group2'),
+                form = call("getData").core.form;
+
+            text1.p.dataParent = '/data/group1';
+            form.moveMug(group1, group2, 'into');
+            util.clickQuestion(text1.p.nodeID);
+            assert(!util.isTreeNodeValid(text1), "text1 should not be valid");
```

Comment:
Glad you mentioned this. Looks like there is a pretty nasty bug here that doesn't allow the data parents to be moved.